packages/mobile/src/fiatExchanges/BidaliScreen.tsx

```diff
@@ -0,0 +1,148 @@
+import colors from '@celo/react-components/styles/colors'
+import { CURRENCIES, CURRENCY_ENUM } from '@celo/utils'
+import { StackScreenProps } from '@react-navigation/stack'
+import React, { useEffect, useMemo, useRef, useState } from 'react'
+import { ActivityIndicator, StyleSheet } from 'react-native'
+import { SafeAreaView } from 'react-native-safe-area-context'
+import { WebView, WebViewMessageEvent } from 'react-native-webview'
+import { useDispatch, useSelector } from 'react-redux'
+import { e164NumberSelector } from 'src/account/selectors'
+import { openUrl } from 'src/app/actions'
+import { bidaliPaymentRequested } from 'src/fiatExchanges/actions'
+import networkConfig from 'src/geth/networkConfig'
+import { celoTokenBalanceSelector } from 'src/goldToken/selectors'
+import i18n from 'src/i18n'
+import { emptyHeader } from 'src/navigator/Headers'
+import { navigate } from 'src/navigator/NavigationService'
+import { Screens } from 'src/navigator/Screens'
+import { TopBarTextButton } from 'src/navigator/TopBarButton'
+import { StackParamList } from 'src/navigator/types'
+import { stableTokenBalanceSelector } from 'src/stableToken/reducer'
+
+function useInitialJavaScript(
+  currency: CURRENCY_ENUM,
+  jsonBalances: string,
+  e164PhoneNumber: string | null
+) {
+  const [initialJavaScript, setInitialJavaScript] = useState<string | null>()
+  useEffect(() => {
+    if (initialJavaScript || !e164PhoneNumber) {
+      return
+    }
+
+    setInitialJavaScript(`
+      window.valora = {
+        paymentCurrency: "${CURRENCIES[currency].code.toUpperCase()}",
+        phoneNumber: "${e164PhoneNumber}",
+        balances: ${jsonBalances},
+        onPaymentRequest: function (paymentRequest) {
+          var payload = { method: 'onPaymentRequest', data: paymentRequest };
+          window.ReactNativeWebView.postMessage(JSON.stringify(payload));
+        },
+        openUrl: function (url) {
+          var payload = { method: 'openUrl', data: { url } };
+          window.ReactNativeWebView.postMessage(JSON.stringify(payload));
+        }
+      };
+      true; // note: this is required, or you'll sometimes get silent failures
+    `)
+  }, [jsonBalances, e164PhoneNumber, initialJavaScript])
+
+  return initialJavaScript
+}
+
+type RouteProps = StackScreenProps<StackParamList, Screens.BidaliScreen>
+type Props = RouteProps
+
+function BidaliScreen({ route }: Props) {
+  const onMessage = (event: WebViewMessageEvent) => {
+    const { method, data } = JSON.parse(event.nativeEvent.data)
+    switch (method) {
+      case 'onPaymentRequest':
+        const { amount, address, currency, description, chargeId } = data
+        const onPaymentSent = () => {
+          webViewRef.current?.injectJavaScript(`window.valora.paymentSent();`)
+        }
+        const onCancelled = () => {
+          webViewRef.current?.injectJavaScript(`window.valora.paymentCancelled();`)
+        }
```

Comment:
Can these be defined outside of the `onMessage` function so that they aren't redefined on every message?

Reply:
They could yes, but I preferred to leave them closer because they are only related to the current request received.